operator/internal/controller/redpanda/redpanda_controller.go

```diff
@@ -112,390 +107,319 @@ type RedpandaReconciler struct {
 
-	// Examine if the object is under deletion
-	if !rp.ObjectMeta.DeletionTimestamp.IsZero() {
-		return ctrl.Result{}, r.reconcileDelete(ctx, rp)
-	}
-
-	if !isRedpandaManaged(ctx, rp) {
-		if controllerutil.ContainsFinalizer(rp, FinalizerKey) {
-			// if no longer managed by us, attempt to remove the finalizer
-			controllerutil.RemoveFinalizer(rp, FinalizerKey)
-			if err := r.Client.Update(ctx, rp); err != nil {
-				return ctrl.Result{}, errors.WithStack(err)
-			}
-		}
-
-		return ctrl.Result{}, nil
-	}
-
+	// TODO: remove this when we delete the managed decommission controller
 	_, ok := rp.GetAnnotations()[resources.ManagedDecommissionAnnotation]
 	if ok {
-		log.V(logger.TraceLevel).Info("Managed decommission")
+		log.V(TraceLevel).Info("Managed decommission")
 		return ctrl.Result{}, nil
 	}
 
-	// add finalizer if not exist
-	if !controllerutil.ContainsFinalizer(rp, FinalizerKey) {
-		patch := client.MergeFrom(rp.DeepCopy())
-		controllerutil.AddFinalizer(rp, FinalizerKey)
-		if err := r.Client.Patch(ctx, rp, patch); err != nil {
-			log.Error(err, "unable to register finalizer")
-			return ctrl.Result{}, errors.WithStack(err)
-		}
-	}
-
-	// Upgrade checks. Don't reconcile if UseFlux is true or if ChartRef is set.
-	if rp.Spec.ChartRef.UseFlux != nil && *rp.Spec.ChartRef.UseFlux {
-		log.Error(nil, "useFlux: true is no longer supported. Please downgrade or unset `useFlux`")
-		return ctrl.Result{}, nil
+	// grab our existing and desired pool resources
+	// so that we can immediately calculate cluster status
+	// from and sync in any subsequent operation that
+	// early returns
+	injectedConfigVersion := ""
+	if rp.Annotations != nil && rp.Annotations[RestartClusterOnConfigChangeKey] == "true" {
+		injectedConfigVersion = rp.Status.ConfigVersion
 	}
-
-	if rp.Spec.ChartRef.ChartVersion != "" {
-		msg := "Specifying chartVersion is no longer supported. Please downgrade or unset `chartRef.chartVersion`"
-		log.Error(nil, msg)
-		r.EventRecorder.Eventf(rp, "Warning", redpandav1alpha2.EventSeverityError, msg)
-		return ctrl.Result{}, nil
+	pools, err := r.LifecycleClient.FetchExistingAndDesiredPools(ctx, rp, injectedConfigVersion)
+	if err != nil {
+		log.Error(err, "fetching pools")
+		return ctrl.Result{}, err
 	}
 
-	if err := r.reconcileResources(ctx, rp); err != nil {
-		return ctrl.Result{}, errors.WithStack(err)
-	}
+	status := lifecycle.NewClusterStatus()
+	status.Pools = pools.PoolStatuses()
 
-	if err := r.reconcileStatus(ctx, rp); err != nil {
-		return ctrl.Result{}, errors.WithStack(err)
+	if pools.AnyReady() {
+		status.Status.SetReady(statuses.ClusterReadyReasonReady)
+	} else {
+		status.Status.SetReady(statuses.ClusterReadyReasonNotReady, "No pods are ready")
 	}
 
-	if err := r.reconcileLicense(ctx, rp); err != nil {
-		return ctrl.Result{}, errors.WithStack(err)
+	// Examine if the object is under deletion
+	if !rp.ObjectMeta.DeletionTimestamp.IsZero() || !isRedpandaManaged(ctx, rp) {
```

Comment:
These two cases are supposed to be distinct are they not? If the repanda instance isn't managed, the operator is supposed to ignore it AFAIK. Treating this as a deletion is terrifying.

Reply:
Ah, yeah, sorry, you're right, I think that this got merged as I was refactoring the finalizer removal code and before I added the cleanup code. The original code was doing:

```
	if !isRedpandaManaged(ctx, rp) {
		if controllerutil.ContainsFinalizer(rp, FinalizerKey) {
			// if no longer managed by us, attempt to remove the finalizer
			controllerutil.RemoveFinalizer(rp, FinalizerKey)
			if err := r.Client.Update(ctx, rp); err != nil {
				return ctrl.Result{}, errors.WithStack(err)
			}
		}

		return ctrl.Result{}, nil
	}
```

Which also seems kind of odd to me... but I can re-split this up (do you think I should also keep the finalizer removal code or just go with a no-op sans removal at that point)?